Sources/PackageCollections/PackageIndexAndCollections.swift

```diff
@@ -0,0 +1,300 @@
+    
+    /// Returns metadata for the package identified by the given `PackageIdentity`, using package index (if configured)
+    /// and collections data.
+    ///
+    /// A failure is returned if the package is not found.
+    ///
+    /// - Parameters:
+    ///   - identity: The package identity
+    ///   - location: The package location (optional for deduplication)
+    ///   - collections: Optional. If specified, only these collections are used to construct the result.
+    ///   - callback: The closure to invoke when result becomes available
+    public func getPackageMetadata(
+        identity: PackageIdentity,
+        location: String? = nil,
+        collections: Set<PackageCollectionsModel.CollectionIdentifier>? = nil,
+        callback: @escaping (Result<PackageCollectionsModel.PackageMetadata, Error>) -> Void
+    ) {
+        // Package index not available - fallback to collections
+        guard self.index.isEnabled else {
+            return self.collections.getPackageMetadata(identity: identity, location: location, collections: collections, callback: callback)
+        }
+                
+        // Get metadata using both package index and collections
+        let sync = DispatchGroup()
+        let results = ThreadSafeKeyValueStore<Source, Result<PackageCollectionsModel.PackageMetadata, Error>>()
+
+        sync.enter()
+        // This uses package index only
+        self.index.getPackageMetadata(identity: identity, location: location) { result in
+            defer { sync.leave() }
+            results[.index] = result
+        }
+
+        sync.enter()
+        // This uses either package index or "alternative" (e.g., GitHub) as metadata provider,
+        // then merge the supplementary metadata with data coming from collections. The package
+        // must belong to at least one collection.
+        self.collections.getPackageMetadata(identity: identity, location: location, collections: collections) { result in
+            defer { sync.leave() }
+            results[.collections] = result
+        }
+        
+        sync.notify(queue: .sharedConcurrent) {
+            guard let indexResult = results[.index], let collectionsResult = results[.collections] else {
+                return callback(.failure(InternalError("Should contain results from package index and collections")))
+            }
+
+            switch indexResult {
+            case .success(let metadataResult):
+                // Metadata from `PackageIndex`
+                callback(.success(
+                    PackageCollectionsModel.PackageMetadata(
+                        package: metadataResult.package,
+                        collections: collectionsResult.success?.collections ?? [],
+                        provider: metadataResult.provider
+                    )
+                ))
+            case .failure(let indexError):
+                switch collectionsResult {
+                case .success(let metadataResult):
```

Comment:
considers switching over both results to avoid the nested switch, e.g.

```
switch (indexResult, collectionsResult) {
case (.success(let metadataResult), _):
 ...
case (.failure(let indexError), .success(let metadataResult)):
...
case (.failure(let indexError), .failure(let collectionsError))::
...
}
```